File: packages/deon-go/interpreter.go
package deon

import (
	"strconv"
	"strings"
)

// interpreter evaluates a parsed document into a Deon value (specification 11). Declarations are
// resolved lazily and memoized, which is equivalent to the topological resolution the specification
// describes and detects a cycle at the reference that closes it (§11.2).
type interpreter struct {
	options      *ParseOptions
	declarations map[string]*declaration
	cache        map[string]Value
	resolving    map[string]bool // declaration names mid-evaluation, for cycle detection
	calling      map[string]bool // entity names mid-call, for recursive-call cycles
	locals       []map[string]string
	opened       map[string]bool // resource identifiers being loaded, for import cycles
}

func newInterpreter(options *ParseOptions) *interpreter {
	return &interpreter{
		options:      options,
		declarations: map[string]*declaration{},
		cache:        map[string]Value{},
		resolving:    map[string]bool{},
		calling:      map[string]bool{},
		opened:       map[string]bool{},
	}
}

func (in *interpreter) run(doc *document) Value {
	// A self-importing document is a cycle, not a loop: seed the open set with this document's name.
	if in.options.sourceName() != "" {
		in.opened[in.options.sourceName()] = true
	}

	for i := range doc.declarations {
		decl := &doc.declarations[i]
		if first, exists := in.declarations[decl.name]; exists {
			// The repeat stays the primary span; the first declaration becomes the single related span
			// the reader is sent to (spec/diagnostics.md).
			fail(DuplicateDeclaration,
				"The name '"+decl.name+"' is declared more than once.", decl.nameSpan, first.nameSpan)
		}
		in.declarations[decl.name] = decl
	}

	return in.eval(doc.root)
}

// #region evaluation
func (in *interpreter) eval(n node) Value {
	switch node := n.(type) {
	case *scalarNode:
		return in.evalScalar(node)
	case *mapNode:
		return in.evalMap(node)
	case *listNode:
		return in.evalList(node)
	case *structureNode:
		return in.evalStructure(node)
	case *linkNode:
		// A link's diagnostic (unresolved, or a cycle it closes) is reported at the link as written,
		// from its `#`, rather than at the reference name after it.
		ref := node.ref
		ref.span = node.span
		return in.resolveReference(ref)
	case *callNode:
		return in.evalCall(node)
	default:
		return ""
	}
}

func (in *interpreter) evalScalar(n *scalarNode) Value {
	var b strings.Builder
	for _, part := range n.parts {
		if part.interp == nil {
			b.WriteString(part.literal)
			continue
		}
		// An interpolation is reported at the string that carries it, not at a position inside it: the
		// reference within was recovered by decoding and has no source position of its own.
		ref := *part.interp
		ref.span = n.span
		value := in.resolveReference(ref)
		text, ok := value.(string)
		if !ok {
			fail(TypeMismatch, "An interpolation must resolve to a string.", n.span)
		}
		b.WriteString(text)
	}
	return b.String()
}

func (in *interpreter) evalMap(n *mapNode) Value {
	result := NewMap()
	for _, entry := range n.entries {
		if entry.spread != nil {
			in.spreadIntoMap(result, *entry.spread)
			continue
		}
		result.Set(entry.key, in.eval(entry.value))
	}
	return result
}

func (in *interpreter) evalList(n *listNode) Value {
	result := []Value{}
	for _, item := range n.items {
		if item.spread != nil {
			result = in.spreadIntoList(result, *item.spread)
			continue
		}
		result = append(result, in.eval(item.value))
	}
	return result
}

func (in *interpreter) evalStructure(n *structureNode) Value {
	result := []Value{}
	for _, row := range n.rows {
		entry := NewMap()
		for i, field := range n.fields {
			entry.Set(field, in.eval(row[i]))
		}
		result = append(result, entry)
	}
	return result
}

// #endregion evaluation

// #region spread
func (in *interpreter) spreadIntoMap(dest *Map, ref reference) {
	value := in.resolveReference(ref)
	switch source := value.(type) {
	case *Map:
		for _, key := range source.Keys() {
			v, _ := source.Get(key)
			dest.Set(key, v)
		}
	case string:
		// A string spreads into a map using decimal character indices (specification 7).
		for i, r := range []rune(source) {
			dest.Set(strconv.Itoa(i), string(r))
		}
	default:
		fail(TypeMismatch, "A list cannot spread into a map.", ref.span)
	}
}

func (in *interpreter) spreadIntoList(dest []Value, ref reference) []Value {
	value := in.resolveReference(ref)
	switch source := value.(type) {
	case []Value:
		return append(dest, source...)
	case string:
		// A string spreads into a list as Unicode code points (specification 7).
		for _, r := range source {
			dest = append(dest, string(r))
		}
		return dest
	default:
		fail(TypeMismatch, "A map cannot spread into a list.", ref.span)
		return dest
	}
}

// #endregion spread

// #region references
func (in *interpreter) resolveReference(ref reference) Value {
	if ref.env {
		if in.options.Environment != nil {
			if value, ok := in.options.Environment[ref.head]; ok {
				return value
			}
		}
		return ""
	}

	value := in.resolveHead(ref.head, ref.span)
	return in.applyAccess(value, ref.access, ref.span)
}

func (in *interpreter) resolveHead(name string, span Span) Value {
	// A call-local binding shadows an outer leaflink for the duration of the call (specification 10).
	for i := len(in.locals) - 1; i >= 0; i-- {
		if value, ok := in.locals[i][name]; ok {
			return value
		}
	}

	decl, ok := in.declarations[name]
	if !ok {
		fail(UnresolvedLink, "There is no declaration named '"+name+"'.", span)
	}

	if cached, done := in.cache[name]; done {
		return cached
	}
	if in.resolving[name] {
		// Reported at the reference that closed the loop, not the declaration that opened it (§11.2).
		fail(Cycle, "The declaration '"+name+"' depends on itself.", span)
	}

	in.resolving[name] = true
	value := in.evalDeclaration(decl)
	delete(in.resolving, name)
	in.cache[name] = value
	return value
}

func (in *interpreter) evalDeclaration(decl *declaration) Value {
	switch decl.kind {
	case declLeaflink:
		return in.eval(decl.value)
	case declImport:
		return in.evalImport(decl)
	case declInject:
		return in.evalInject(decl)
	default:
		return ""
	}
}

func (in *interpreter) applyAccess(value Value, access []accessSegment, span Span) Value {
	for _, segment := range access {
		switch container := value.(type) {
		case *Map:
			key := segment.name
			member, ok := container.Get(key)
			if !ok {
				fail(UnresolvedLink, "There is no member '"+key+"'.", span)
			}
			value = member
		case []Value:
			if !segment.byIndex {
				fail(UnresolvedLink, "A list is indexed by a number, not by '"+segment.name+"'.", span)
			}
			if segment.index < 0 || segment.index >= len(container) {
				fail(UnresolvedLink, "The list index "+strconv.Itoa(segment.index)+" is out of range.", span)
			}
			value = container[segment.index]
		default:
			fail(UnresolvedLink, "A string has no members to access.", span)
		}
	}
	return value
}

// #endregion references

// #region entity calls
func (in *interpreter) evalCall(n *callNode) Value {
	name := n.ref.head
	if len(n.ref.access) > 0 {
		fail(UnresolvedLink, "A call names a leaflink directly.", n.span)
	}

	decl, ok := in.declarations[name]
	if !ok || decl.kind != declLeaflink {
		fail(UnresolvedLink, "There is no entity named '"+name+"' to call.", n.span)
	}

	params := interpolationNames(decl.value)

	locals := map[string]string{}
	for _, arg := range n.args {
		if _, dup := locals[arg.name]; dup {
			// Every argument fault anchors its primary span at the opening '(' (argsSpan); the offending
			// argument's name becomes the single related span the reader is sent to (spec §11.2).
			fail(EntityArgument, "The argument '"+arg.name+"' is given more than once.", n.argsSpan, arg.nameSpan)
		}
		if !params[arg.name] {
			fail(EntityArgument, "'"+name+"' has no parameter '"+arg.name+"'.", n.argsSpan, arg.nameSpan)
		}
		value := in.eval(arg.value)
		text, isString := value.(string)
		if !isString {
			fail(EntityArgument, "The argument '"+arg.name+"' must be a string.", n.argsSpan, arg.nameSpan)
		}
		locals[arg.name] = text
	}
	for param := range params {
		if _, given := locals[param]; !given {
			fail(EntityArgument, "'"+name+"' is missing the argument '"+param+"'.", n.argsSpan)
		}
	}

	if in.calling[name] {
		fail(Cycle, "The entity '"+name+"' calls itself.", n.span)
	}
	in.calling[name] = true
	in.locals = append(in.locals, locals)

	// Every call evaluates an independent copy: the entity body is re-evaluated from its node, so two
	// calls with different arguments do not share a memoized value.
	value := in.eval(decl.value)

	in.locals = in.locals[:len(in.locals)-1]
	delete(in.calling, name)
	return value
}

// interpolationNames collects the interpolation names an entity carries, which are its exact
// parameter set (specification 10). A link (`#voice`) composes and stays private; an interpolation
// (`#{voice}`) is a hole and is always a parameter, even when a leaflink of that name exists.
func interpolationNames(n node) map[string]bool {
	names := map[string]bool{}
	var walk func(node)
	walk = func(n node) {
		switch node := n.(type) {
		case *scalarNode:
			for _, part := range node.parts {
				if part.interp != nil && !part.interp.env {
					names[part.interp.head] = true
				}
			}
		case *mapNode:
			for _, entry := range node.entries {
				if entry.value != nil {
					walk(entry.value)
				}
			}
		case *listNode:
			for _, item := range node.items {
				if item.value != nil {
					walk(item.value)
				}
			}
		case *structureNode:
			for _, row := range node.rows {
				for _, cell := range row {
					walk(cell)
				}
			}
		}
	}
	walk(n)
	return names
}

// #endregion entity calls

File: packages/deon-java/source/deon/Interpreter.java
package deon;

import java.io.IOException;
import java.nio.charset.CodingErrorAction;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.function.Supplier;

/**
 * Evaluating a parsed document into a Deon value (specification 11). Declarations resolve lazily and are
 * memoized, which is equivalent to the topological resolution the specification describes and detects a
 * cycle at the reference that closes it. A diagnostic is thrown; an import re-anchors any diagnostic it
 * raises to its own statement, so a fault inside a resource is reported at the line that imported it.
 */
final class Interpreter {
    private final ParseOptions options;
    private final Map<String, Declaration> declarations = new LinkedHashMap<>();
    private final Map<String, Object> cache = new HashMap<>();
    private final Set<String> resolving = new HashSet<>();
    private final Set<String> calling = new HashSet<>();
    private final Deque<Map<String, String>> locals = new ArrayDeque<>();
    private final Set<String> opened;
    private String sourceName;
    private String filebase;

    Interpreter(ParseOptions options, Set<String> opened, String sourceName, String filebase) {
        this.options = options;
        this.opened = opened;
        this.sourceName = sourceName;
        this.filebase = filebase;
    }

    void register(List<Declaration> declarationList) {
        for (Declaration d : declarationList) {
            if (declarations.containsKey(d.name)) {
                throw new DeonException(Code.DUPLICATE_DECLARATION, "The name is declared more than once.", d.nameSpan);
            }
            declarations.put(d.name, d);
        }
    }

    Object run(Node root) {
        return eval(root);
    }

    // #region evaluation
    private Object eval(Node n) {
        if (n instanceof ScalarNode scalar) {
            return evalScalar(scalar);
        }
        if (n instanceof MapNode map) {
            return evalMap(map);
        }
        if (n instanceof ListNode list) {
            return evalList(list);
        }
        if (n instanceof StructureNode structure) {
            return evalStructure(structure);
        }
        if (n instanceof LinkNode link) {
            Reference ref = link.ref;
            ref.span = n.span(); // a link's diagnostic is at its #, not the name after it
            return resolveReference(ref);
        }
        return evalCall((CallNode) n);
    }

    private Object evalScalar(ScalarNode n) {
        StringBuilder b = new StringBuilder();
        for (StringPart part : n.parts) {
            if (!part.isInterp) {
                b.append(part.literal);
                continue;
            }
            // An interpolation is reported at the string that carries it, not inside it: the reference
            // was recovered by decoding and has no source position of its own.
            Reference ref = part.interp;
            ref.span = n.span;
            Object value = resolveReference(ref);
            if (!(value instanceof String s)) {
                throw new DeonException(Code.TYPE_MISMATCH, "An interpolation must resolve to a string.", n.span);
            }
            b.append(s);
        }
        return b.toString();
    }

    private void spreadIntoMap(DeonMap dest, Reference ref) {
        Object value = resolveReference(ref);
        if (value instanceof DeonMap map) {
            for (String key : map.keys()) {
                dest.set(key, map.get(key));
            }
        } else if (value instanceof String s) {
            // A string spreads into a map using decimal character indices (section 7).
            int index = 0;
            for (int i = 0; i < s.length(); ) {
                int cp = s.codePointAt(i);
                dest.set(Integer.toString(index), new String(Character.toChars(cp)));
                i += Character.charCount(cp);
                index++;
            }
        } else {
            throw new DeonException(Code.TYPE_MISMATCH, "A list cannot spread into a map.", ref.span);
        }
    }

    private void spreadIntoList(List<Object> dest, Reference ref) {
        Object value = resolveReference(ref);
        if (value instanceof List<?> list) {
            dest.addAll(list);
        } else if (value instanceof String s) {
            for (int i = 0; i < s.length(); ) {
                int cp = s.codePointAt(i);
                dest.add(new String(Character.toChars(cp)));
                i += Character.charCount(cp);
            }
        } else {
            throw new DeonException(Code.TYPE_MISMATCH, "A map cannot spread into a list.", ref.span);
        }
    }

    private Object evalMap(MapNode n) {
        DeonMap result = new DeonMap();
        for (MapEntry e : n.entries) {
            if (e.spread != null) {
                spreadIntoMap(result, e.spread);
                continue;
            }
            result.set(e.key, eval(e.value));
        }
        return result;
    }

    private Object evalList(ListNode n) {
        List<Object> result = new ArrayList<>();
        for (ListItem it : n.items) {
            if (it.spread != null) {
                spreadIntoList(result, it.spread);
                continue;
            }
            result.add(eval(it.value));
        }
        return result;
    }

    private Object evalStructure(StructureNode n) {
        List<Object> result = new ArrayList<>();
        for (List<Node> row : n.rows) {
            DeonMap entry = new DeonMap();
            for (int c = 0; c < n.fields.size(); c++) {
                entry.set(n.fields.get(c), eval(row.get(c)));
            }
            result.add(entry);
        }
        return result;
    }

    private Object evalCall(CallNode n) {
        String name = n.ref.head;
        if (!n.ref.access.isEmpty()) {
            throw new DeonException(Code.UNRESOLVED_LINK, "A call names a leaflink directly.", n.span);
        }
        Declaration decl = declarations.get(name);
        if (decl == null || decl.kind != Declaration.Kind.LEAFLINK) {
            throw new DeonException(Code.UNRESOLVED_LINK, "There is no entity to call.", n.span);
        }

        Set<String> params = interpolationNames(decl.value);
        Map<String, String> bindings = new LinkedHashMap<>();
        for (CallArg a : n.args) {
            if (bindings.containsKey(a.name)) {
                throw new DeonException(Code.ENTITY_ARGUMENT, "An argument is given more than once.", a.nameSpan);
            }
            if (!params.contains(a.name)) {
                throw new DeonException(Code.ENTITY_ARGUMENT, "There is no such parameter.", n.argsSpan);
            }
            Object v = eval(a.value);
            if (!(v instanceof String s)) {
                throw new DeonException(Code.ENTITY_ARGUMENT, "An argument must be a string.", a.nameSpan);
            }
            bindings.put(a.name, s);
        }
        for (String param : params) {
            if (!bindings.containsKey(param)) {
                throw new DeonException(Code.ENTITY_ARGUMENT, "A required argument is missing.", n.argsSpan);
            }
        }

        if (calling.contains(name)) {
            throw new DeonException(Code.CYCLE, "The entity calls itself.", n.span);
        }
        calling.add(name);
        locals.push(bindings);
        try {
            return eval(decl.value);
        } finally {
            locals.pop();
            calling.remove(name);
        }
    }
    // #endregion

    // #region references
    private Object applyAccess(Object value, List<AccessSeg> access, Span span) {
        for (AccessSeg seg : access) {
            if (value instanceof DeonMap map) {
                if (!map.has(seg.name)) {
                    throw new DeonException(Code.UNRESOLVED_LINK, "There is no such member.", span);
                }
                value = map.get(seg.name);
            } else if (value instanceof List<?> list) {
                if (!seg.byIndex) {
                    throw new DeonException(Code.UNRESOLVED_LINK, "A list is indexed by a number.", span);
                }
                if (seg.index < 0 || seg.index >= list.size()) {
                    throw new DeonException(Code.UNRESOLVED_LINK, "The list index is out of range.", span);
                }
                value = list.get(seg.index);
            } else {
                throw new DeonException(Code.UNRESOLVED_LINK, "A string has no members to access.", span);
            }
        }
        return value;
    }

    private Object evalDeclaration(Declaration decl) {
        return switch (decl.kind) {
            case LEAFLINK -> eval(decl.value);
            case IMPORT -> evalImport(decl);
            case INJECT -> evalInject(decl);
        };
    }

    private Object resolveHead(String name, Span span) {
        for (Map<String, String> frame : locals) { // top of stack first
            if (frame.containsKey(name)) {
                return frame.get(name);
            }
        }
        Declaration decl = declarations.get(name);
        if (decl == null) {
            throw new DeonException(Code.UNRESOLVED_LINK, "There is no such declaration.", span);
        }
        if (cache.containsKey(name)) {
            return cache.get(name);
        }
        if (resolving.contains(name)) {
            throw new DeonException(Code.CYCLE, "The declaration depends on itself.", span);
        }
        resolving.add(name);
        Object value;
        try {
            value = evalDeclaration(decl);
        } finally {
            resolving.remove(name);
        }
        cache.put(name, value);
        return value;
    }

    private Object resolveReference(Reference ref) {
        if (ref.env) {
            return options.environment.getOrDefault(ref.head, "");
        }
        Object value = resolveHead(ref.head, ref.span);
        return applyAccess(value, ref.access, ref.span);
    }
    // #endregion

    // #region entity parameters
    /** The interpolation names of an entity body: its exact parameter set (specification 10). */
    static Set<String> interpolationNames(Node n) {
        Set<String> set = new LinkedHashSet<>();
        collectInterpNames(n, set);
        return set;
    }

    private static void collectInterpNames(Node n, Set<String> set) {
        if (n instanceof ScalarNode scalar) {
            for (StringPart part : scalar.parts) {
                if (part.isInterp && !part.interp.env) {
                    set.add(part.interp.head);
                }
            }
        } else if (n instanceof MapNode map) {
            for (MapEntry e : map.entries) {
                if (e.value != null) {
                    collectInterpNames(e.value, set);
                }
            }
        } else if (n instanceof ListNode list) {
            for (ListItem it : list.items) {
                if (it.value != null) {
                    collectInterpNames(it.value, set);
                }
            }
        } else if (n instanceof StructureNode structure) {
            for (List<Node> row : structure.rows) {
                for (Node cell : row) {
                    collectInterpNames(cell, set);
                }
            }
        }
    }
    // #endregion

    // #region path and URL helpers
    static boolean isUrl(String t) {
        return t.startsWith("http://") || t.startsWith("https://");
    }

    /** The trailing ".ext" of the last path segment, or "" if none. */
    static String extensionOf(String t) {
        String path = t;
        if (isUrl(t)) {
            int scheme = t.indexOf("://");
            int slash = t.indexOf('/', scheme + 3);
            path = slash < 0 ? "" : t.substring(slash);
            // The query and fragment are not part of the path: the format is read from the path
            // alone (section 9), so `data.json?v=2` is JSON, not the extension `.json?v=2`.
            int query = path.indexOf('?');
            int fragment = path.indexOf('#');
            int cut = path.length();
            if (query >= 0) {
                cut = query;
            }
            if (fragment >= 0 && fragment < cut) {
                cut = fragment;
            }
            path = path.substring(0, cut);
        }
        int lastSlash = path.lastIndexOf('/');
        String seg = lastSlash < 0 ? path : path.substring(lastSlash + 1);
        int dot = seg.lastIndexOf('.');
        if (dot <= 0) {
            return "";
        }
        return seg.substring(dot);
    }

    /** Collapse "." and ".." segments in an absolute or relative path. */
    static String normalizePath(String path) {
        boolean absolute = !path.isEmpty() && path.charAt(0) == '/';
        Deque<String> segs = new ArrayDeque<>();
        for (String seg : path.split("/")) {
            if (seg.isEmpty() || seg.equals(".")) {
                continue;
            }
            if (seg.equals("..")) {
                // Above the base there is nothing to pop, so a relative path keeps the `..` it
                // cannot resolve (section 9): `a/../../b` normalizes to `../b`, not `b`. A rooted
                // path instead drops what it cannot climb above its root.
                if (!segs.isEmpty() && !segs.peekLast().equals("..")) {
                    segs.removeLast();
                } else if (!absolute) {
                    segs.addLast("..");
                }
                continue;
            }
            segs.addLast(seg);
        }
        String joined = String.join("/", segs);
        if (absolute) {
            return "/" + joined;
        }
        return joined.isEmpty() ? "." : joined;
    }

    static String directoryOf(String t) {
        if (isUrl(t)) {
            int scheme = t.indexOf("://");
            int slash = t.indexOf('/', scheme + 3);
            if (slash < 0) {
                return t;
            }
            int last = t.lastIndexOf('/');
            return t.substring(0, last + 1); // include trailing slash
        }
        int last = t.lastIndexOf('/');
        if (last < 0) {
            return ".";
        }
        if (last == 0) {
            return "/";
        }
        return t.substring(0, last);
    }

    static String pathJoin(String base, String rel) {
        if (base.isEmpty()) {
            return normalizePath(rel);
        }
        return normalizePath(base + "/" + rel);
    }

    static String urlJoin(String base, String rel) {
        if (isUrl(rel)) {
            return rel;
        }
        int scheme = base.indexOf("://");
        int pathStart = base.indexOf('/', scheme + 3);
        int prefixLen = pathStart < 0 ? base.length() : pathStart;
        String basePath = pathStart < 0 ? "/" : base.substring(pathStart);

        String combined;
        if (rel.startsWith("/")) {
            combined = normalizePath(rel);
        } else {
            int lastSlash = basePath.lastIndexOf('/');
            String dir = lastSlash < 0 ? "" : basePath.substring(0, lastSlash + 1);
            combined = normalizePath(dir + rel);
        }
        return base.substring(0, prefixLen) + combined;
    }

    private String mapAbsolute(String target) {
        Map<String, String> abs = options.absolutePaths;
        if (abs.isEmpty()) {
            return target;
        }
        if (abs.containsKey(target)) {
            return abs.get(target);
        }
        String bestKey = null;
        int bestLen = 0;
        for (String key : abs.keySet()) {
            if (key.length() < 2 || !key.endsWith("/*")) {
                continue;
            }
            String prefix = key.substring(0, key.length() - 1); // keep trailing slash
            if (target.startsWith(prefix) && prefix.length() > bestLen) {
                bestKey = key;
                bestLen = prefix.length();
            }
        }
        if (bestKey == null) {
            return target;
        }
        String mapped = abs.get(bestKey);
        String remainder = target.substring(bestLen);
        while (mapped.endsWith("/")) {
            mapped = mapped.substring(0, mapped.length() - 1);
        }
        return mapped + "/" + remainder;
    }

    private String resolveTarget(String target) {
        String resolved = target;
        if (isUrl(target)) {
            // already absolute
        } else if (target.startsWith("/")) {
            // logical absolute, left for the absolutePaths mapping
        } else if (filebase != null && !filebase.isEmpty() && isUrl(filebase)) {
            String base = filebase.endsWith("/") ? filebase : filebase + "/";
            resolved = urlJoin(base, target);
        } else {
            resolved = pathJoin(filebase == null ? "" : filebase, target);
        }
        return mapAbsolute(resolved);
    }

    private static String importTarget(String target) {
        if (extensionOf(target).isEmpty()) {
            return target + ".deon";
        }
        return target;
    }
    // #endregion

    // #region loading
    private record Fetched(String data, String filetype, String filebase, String resourceId) {
    }

    private String tokenFor(String target) {
        if (!isUrl(target)) {
            return "";
        }
        int scheme = target.indexOf("://");
        int start = scheme + 3;
        int end = start;
        while (end < target.length() && target.charAt(end) != '/' && target.charAt(end) != ':') {
            end++;
        }
        String host = target.substring(start, end).toLowerCase();
        return options.authorization.getOrDefault(host, "");
    }

    private Fetched loadResource(String target, String kind, String token, Span span) {
        String filetype = kind.equals("import") ? extensionOf(target) : "";
        String base = directoryOf(target);

        String mem = options.resources.get(target);
        if (mem != null) {
            return new Fetched(mem, filetype, base, target);
        }

        if (isUrl(target)) {
            if (!options.allowNetwork) {
                throw new DeonException(Code.CAPABILITY_DENIED, "The resource was not permitted: network access is not allowed.", span);
            }
            String credential = token != null && !token.isEmpty() ? token : tokenFor(target);
            String cached = Cache.read(options, target, credential);
            if (cached != null) {
                return new Fetched(cached, filetype, base, target);
            }
            String body = Network.httpGet(target, kind, credential, span);
            Cache.write(options, target, credential, body);
            return new Fetched(body, filetype, base, target);
        }

        if (!options.allowFilesystem) {
            throw new DeonException(Code.CAPABILITY_DENIED, "The resource was not permitted: filesystem access is not allowed.", span);
        }
        byte[] bytes;
        try {
            bytes = Files.readAllBytes(Path.of(target));
        } catch (IOException e) {
            throw new DeonException(Code.RESOURCE_IO, "Unable to read the resource.", span);
        }
        if (!isValidUtf8(bytes)) {
            throw new DeonException(Code.RESOURCE_FORMAT, "The resource is not valid UTF-8.", span);
        }
        return new Fetched(new String(bytes, StandardCharsets.UTF_8), filetype, base, target);
    }

    static boolean isValidUtf8(byte[] bytes) {
        var decoder = StandardCharsets.UTF_8.newDecoder()
                .onMalformedInput(CodingErrorAction.REPORT)
                .onUnmappableCharacter(CodingErrorAction.REPORT);
        try {
            decoder.decode(java.nio.ByteBuffer.wrap(bytes));
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    private String resolveAuthenticator(Declaration d) {
        if (d.authenticator == null) {
            return "";
        }
        Object v = eval(d.authenticator);
        if (!(v instanceof String s)) {
            throw new DeonException(Code.TYPE_MISMATCH, "An authenticator must resolve to a string.", d.span);
        }
        return s;
    }

    // Run a sub-evaluation with a boundary that re-anchors any diagnostic to `at` — unless it is a
    // cycle, which keeps its own span.
    private Object reanchored(Supplier<Object> body, Span at) {
        try {
            return body.get();
        } catch (DeonException e) {
            if (e.code == Code.CYCLE) {
                throw e;
            }
            Diagnostic primary = e.primary();
            throw new DeonException(e.code, List.of(new Diagnostic(e.code, primary.message, at)));
        }
    }

    private Object evalImport(Declaration d) {
        String token = resolveAuthenticator(d);
        String target = importTarget(resolveTarget(d.target));
        Fetched f = loadResource(target, "import", token, d.span);

        if (opened.contains(f.resourceId())) {
            throw new DeonException(Code.CYCLE, "The resource imports itself.", d.span);
        }
        opened.add(f.resourceId());
        try {
            String ext = f.filetype();
            if (ext.equals(".json")) {
                return reanchored(() -> JsonReader.read(f.data(), Span.head(f.resourceId())), d.span);
            }
            if (!ext.isEmpty() && !ext.equals(".deon")) {
                throw new DeonException(Code.RESOURCE_FORMAT, "The import has an unsupported extension.", d.span);
            }
            return reanchored(() -> importDeonBody(f), d.span);
        } finally {
            opened.remove(f.resourceId());
        }
    }

    private Object importDeonBody(Fetched f) {
        DocumentAst doc = new Parser(f.data(), f.resourceId()).parseDocument();
        Interpreter sub = new Interpreter(options, opened, f.resourceId(), f.filebase());
        sub.register(doc.declarations);
        return sub.eval(doc.root);
    }

    private Object evalInject(Declaration d) {
        String token = resolveAuthenticator(d);
        String target = resolveTarget(d.target);
        Fetched f = loadResource(target, "inject", token, d.span);
        return f.data();
    }
    // #endregion

    // #region entry
    static Object evaluate(DocumentAst doc, ParseOptions options) {
        String sourceName = options.sourceName() == null || options.sourceName().isEmpty() ? "<memory>" : options.sourceName();
        String filebase = options.filebase == null ? "" : options.filebase;
        Set<String> opened = new HashSet<>();
        if (!sourceName.isEmpty()) {
            opened.add(sourceName);
        }
        Interpreter in = new Interpreter(options, opened, sourceName, filebase);
        in.register(doc.declarations);
        return in.eval(doc.root);
    }
    // #endregion
}
